Mentions api

adrianq opened the pull request.

adrianq commented:
Closes #18

tokland commented on dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/InterpretationController.java:
```diff
@@ -342,10 +407,16 @@ public void updateComment( @PathVariable( "uid" ) String uid, @PathVariable( "cu
         }
 
-        interpretationService.updateInterpretation( interpretation );
+//        interpretationService.updateInterpretation( interpretation );
```
Minor: Can we remov the line altogether?

adrianq replied:
Removed!

tokland commented on dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/InterpretationController.java:
```diff
@@ -342,10 +407,16 @@ public void updateComment( @PathVariable( "uid" ) String uid, @PathVariable( "cu
                 }
 
                 comment.setText( content );
+                Set<User> users = interpretationService.getMentionedUsers( content ); 
```
Before you commented that about those 3 instances of getMentionedUsers + setMentions + updateSharingForMentions + save/update + sendNotifications. I guess that to be abstracted we should have an interface for interpretation/comment (mentionable) so we can pass it as an argument. Also, it will require a check of the obj.getId() to see if it should be updated or saved. Anyway, I also don't see it as a priority.

adrianq replied:
Yes, I agree this could be an option. They have a similar approach for "interpretation objects" (reports, charts, etc). An abstract class with some common functionality. Let's keep it in mind and maybe we can add it in the next iteration.

adrianq commented on dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java:
```diff
@@ -107,13 +142,18 @@ public int saveInterpretation( Interpretation interpretation )
+            interpretation.setMentions(users);
         }
 
+        this.updateSharingForMentions(interpretation, users);
```
I moved it inside the if and I also added an extra validation for the sendNotification method

adrianq commented on dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/interpretation/impl/DefaultInterpretationService.java:
```diff
@@ -157,25 +203,96 @@ public void deleteInterpretation( Interpretation interpretation )
+        else {
+            messageContent = new StringBuilder( "You were mentioned in the following interpretation: \n\n" )
+                .append( interpretation.getText() );
+                
```
Removed!

adrianq commented:
I tackled all your comments. Something else that we talked about but I forgot to commit: I removed this [hardcoded read permission string with a proper constant](https://github.com/EyeSeeTea/dhis2-core/pull/23/commits/b4c230a5525baff063bb544212c15e2d772fb075#diff-12fe023f400f7c4d6bf945260269b5d7R270) that was already in place.
I also updated it with the latest from upstream/master
